[1.20.x] More buildscript clean-up

PaintNinja opened the pull request:
Dev scenario (some stuff cached)
Before: https://scans.gradle.com/s/v56a57ffzlnbw (1m 24s)
After: https://scans.gradle.com/s/77ndslyrqrmik (1m 11s)

CI scenario (`gradlew clean` ran first)
Before: https://scans.gradle.com/s/yu2klxvlx2nos (2m 5s)
After: https://scans.gradle.com/s/masboziv56uam (1m 41s)

This PR does various misc. buildscript cleanup, mostly focused on the Groovy side of things. IDE support is improved and performance is a bit better, too. Made as a PR rather than a direct commit to verify it works on CI first and to allow it to be merged alongside something else in one go.

- More lazy task registration and configuration
- Sync buildSrc deps with build deps
- More types and static compilation for better performance and IDE support
- Replace more ext entries with typed local variables
- Do maven artifact checkExists in parallel and use HTTPClient
- Avoid expensive dynamic Groovy method calls for basic null checks, reduce boxing
- Avoid loops for single items and other minor stuff

Known side-effects (noting in case it matters):
- ~~Ordering of launcher json entries is no longer consistent~~ Fixed

PaintNinja committed:
Buildscript clean-up

PaintNinja committed:
Fix CheckExcs

Whoopsie daisy

PaintNinja committed:
Fix typos in credits

PaintNinja committed:
Merge branch 'MinecraftForge:1.20.x' into 1.20.4-groovy-buildscript-cleanup

LexManos commented:
> Ordering of launcher json entries is no longer consistent

This should be a simple fix by using a LinkedHashMap for the json. Its kinda import that this is ordered for both reproducibility and for people reading the json.

PaintNinja committed:
Preserve ordering of libraries in launcher json

- HTTP check and sha1 still done in parallel
- Insertions done sequentially and in an order-preserving manner
- Use `List.addAll()` instead of a loop to improve performance

PaintNinja committed:
Fix method not found in Util#getArtifacts

PaintNinja committed:
Merge branch 'MinecraftForge:1.20.x' into 1.20.4-groovy-buildscript-cleanup

PaintNinja committed:
Fix typo in Util#getArtifacts
